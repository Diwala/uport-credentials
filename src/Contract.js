const arrayContainsArray = require('ethjs-util').arrayContainsArray;

// A derivative work of Nick Dodson's eths-contract https://github.com/ethjs/ethjs-contract/blob/master/src/index.js
const hasTransactionObject = (args) => {
  const txObjectProperties = ['from', 'to', 'data', 'value', 'gasPrice', 'gas'];
  if (typeof args === 'object' && Array.isArray(args) === true && args.length > 0) {
    if (typeof args[args.length - 1] === 'object'
      && (Object.keys(args[args.length - 1]).length === 0
      || arrayContainsArray(Object.keys(args[args.length - 1]), txObjectProperties, true))) {
      return true;
    }
  }

  return false;
}

const getCallableMethodsFromABI = (contractABI) => {
  return contractABI.filter((json) => ((json.type === 'function' || json.type === 'event') && json.name.length > 0));
}

const encodeMethodReadable = (methodObject, methodArgs) => {
  let dataString = `${methodObject.name}(`

  for (let i = 0; i < methodObject.inputs.length; i++) {
    const input = methodObject.inputs[i]
    let argString = `${input.type} `

    if (input.type === 'string') {
      argString += `'${methodArgs[i]}'`
    } else {
      argString += `${methodArgs[i]}`
    }

    dataString += argString

    if ((methodObject.inputs.length - 1) !== i) {
      dataString += `, `
    }
  }
  return dataString += `)`
}

/**
 * A factory function that returns web3-style contract classes, which calls
 * the provided sendTransaction function on the transaction object for each
 * contract method call.  Generally this is used to create signed JWTs from 
 * method calls which can be signed by the uPort mobile app
 *
 * @param {Function} sendTransaction    The function to be called with the transaction
 *                                      object and a method-specific id on every method call
 * @param {Object}   [provider]         A web3 provider to handle requests etc. that don't
 *                                      require sending an ethereum transaction
 *
 * @returns {UportContract Constructor} The web3-style contract, configured with the above
 */
function ContractFactory (sendTransaction) {
  // const Web3Contract = provider && (new Web3(provider)).eth.Contract

  /**
   * The Contract class, which mocks web3 functionality by creating signed transaction
   * requests, which can be passed to the Uport Mobile app to actually send
   */
  class UportContract {
    constructor (jsonInterface = [], address = '0x', options = {}) {      
      // Create a parallel contract using the provider, for handling non-transaction methods
      // this.contract = provider 
      //   ? new Web3Contract(jsonInterface, address, options)
      //   : {}

      this.jsonInterface = jsonInterface
      this.address = address
      this.options = { ...options, address, jsonInterface }

      this.methods = {}
      // this.events = this.contract.events

      // Create a transaction request function for each Contract method
      getCallableMethodsFromABI(jsonInterface).map(({name, type, inputs, constant}) => {
        if (type === 'function') {          
          // Callable functions are available in this.methods[methodName]
          this.methods[name] = (...args) => ({
            // Borrow estimateGas and encodeABI from web3 contract
            // ...(this.contract && this.contract.methods[name](args)),
            // Overwrite the send function with the uPort sendTransaction or request creator 
            send: (txObject) => {
              txObject = {
                ...txObject, 
                to: address, 
                fn: encodeMethodReadable({name, inputs}, args)
              }

              return sendTransaction 
                ? sendTransaction(txObject, txObject.fn) // TODO: Come up with better id
                : txObject
            },
          })
        } else if (constant) {
          this.methods[name] = () => ({
            call: () => { throw new Error('Constant methods do not require uport interaction') }
          })
          // this.methods[name] = provider 
          //   ? this.contract.methods[name]
          //   : () => new Error('Constant functions cannot be computed without a web3 provider')
        } else {
          // What about the fallback function ?? 
          console.warn(`Unhandled ABI method: ${name}`)
        }
      })

      // Wrap other web3 contract api methods
      // this.getPastEvents = this.contract.getPastEvents
      // this.once = this.contract.once
    }

    clone () {
      return new UportContract(this.jsonInterface, this.address, this.options)
    }

    deploy () {
      throw new Error('Contract deployment not yet supported, use web3')
    }
  }

  return UportContract  
}

const buildRequestURI = (txObject) => {
  return `me.uport:${txObject.to}?function=${txObject.fn}`
}

const Contract = ContractFactory(buildRequestURI)

export { ContractFactory, Contract }
