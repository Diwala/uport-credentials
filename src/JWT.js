import { createUnsignedToken, TokenVerifier, decodeToken } from 'jsontokens' 

const JOSE_HEADER = {typ: 'JWT', alg: 'ES256K'}


export function createJWT ({address, signer}, payload) {
  const signingInput = createUnsignedToken(
    JOSE_HEADER,
    {...payload, iss: address, iat: new Date().getTime()}
  )
  return new Promise((resolve, reject) =>
    signer(signingInput, (error, signature) => {
      if (error) return reject(error)
      resolve([signingInput, signature].join('.'))
    })
  )
}

export function verifyJWT ({registry, address}, jwt) {
  return new Promise((resolve, reject) => {
    const {payload} = decodeToken(jwt)
    registry(payload.iss).then(profile => {
      if (!profile) return reject(new Error('No profile found, unable to verify JWT'))
      const publicKey = profile.publicKey.match(/^0x/) ? profile.publicKey.slice(2) : profile.publicKey
      const verifier = new TokenVerifier('ES256K', publicKey)
      if (verifier.verify(jwt)) {
        if (payload.exp && payload.exp <= new Date().getTime()) {
          return reject(new Error('JWT has expired'))
        }
        if (payload.aud && payload.aud !== address) {
          return reject(new Error('JWT audience does not match your address'))
        }
        resolve({payload, profile})
      } else {
        return reject(new Error('Signature invalid for JWT'))
      }
    }).catch(reject)
  })
}
