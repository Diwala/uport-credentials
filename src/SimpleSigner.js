
export default function SimpleSigner(privateKey) {
  return function (data, callback) {
    // TODO implement signing
  }
}