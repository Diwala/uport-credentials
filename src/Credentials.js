import { createJWT, verifyJWT } from './JWT'
import UportLite from 'uport-lite'
import nets from 'nets'
import { isMNID, encode, decode } from 'mnid'

const networks = {
  'mainnet':   {  id: '0x1',
                  registry: '0xab5c8051b9a1df1aab0149f8b0630848b7ecabf6',
                  rpcUrl: 'https://mainnet.infura.io' },
  'ropsten':   {  id: '0x3',
                  registry: '0x41566e3a081f5032bdcad470adb797635ddfe1f0',
                  rpcUrl: 'https://ropsten.infura.io' },
  'kovan':     {  id: '0x2a',
                  registry: '0x5f8e9351dc2d238fb878b6ae43aa740d62fc9758',
                  rpcUrl: 'https://kovan.infura.io' }
  // 'infuranet': {  id: '0x2a'
  //                 registry: '',
  //                 rpcUrl: 'https://infuranet.infura.io' }
}

const DEFAULTNETWORK = 'kovan'

const configNetwork = (net = DEFAULTNETWORK) => {
  if (typeof net === 'object') {
    ['id', 'registry', 'rpcUrl'].forEach((key) => {
      if (!net.hasOwnProperty(key)) throw new Error(`Malformed network config object, object must have '${key}' key specified.`)
    })
    return net
  } else if (typeof net === 'string') {
    if (!networks[net]) throw new Error(`Network configuration not available for '${net}'`)
    return networks[net]
  }

  throw new Error(`Network configuration object or network string required`)
}

export default class Credentials {
  constructor (settings = {}) {
  this.settings = settings
  this.settings.network = configNetwork(settings.network)
  this.net = this.settings.network
  if (settings.address) {
    this.settings.address = isMNID(settings.address) ? settings.address : encode({network: this.net.id, address: settings.address})
  }
  if (!this.settings.registry) {
    const registryNetwork = {[this.net.id]: {registry: this.net.registry, rpcUrl: this.net.rpcUrl}}
    const registry = UportLite({networks: registryNetwork})
    this.settings.registry = (address) => new Promise((resolve, reject) => {
      const addressMNID = isMNID(address) ? address : encode({network: this.net.id, address})
      registry(address, (error, profile) => {
        if (error) return reject(error)
        resolve(profile)
      })
    })
  }
}

  // Create request token
  createRequest (params = {}) {
    const payload = {}
    if (params.requested) {
      payload.requested = params.requested
    }
    if (params.verified) {
      payload.verified = params.verified
    }
    if (params.notifications) {
      payload.permissions = ['notifications']
    }
    if (params.callbackUrl) {
      payload.callback = params.callbackUrl
    }
    return createJWT(this.settings, {...payload, type: 'shareReq'})
  }

  // Receive response token from user and return data to promise
  receive (token, callbackUrl = null) {
    return verifyJWT(this.settings, token, callbackUrl).then(({payload, profile}) => {
      const credentials = {...profile, ...(payload.own || {}), ...(payload.capabilities && payload.capabilities.length === 1 ? {pushToken: payload.capabilities[0]} : {}), address: payload.iss}
      if (payload.verified) {
        return Promise.all(payload.verified.map(token => verifyJWT(this.settings, token))).then(verified => {
          return {...credentials, verified: verified.map(v => ({...v.payload, jwt: v.jwt}))}
        })
      } else {
        return credentials
      }
    })
  }

  push (token, {url}) {
    return new Promise((resolve, reject) => {
      if (!token) {
        return reject(new Error('Missing push notification token'))
      }
      if (!url) {
        return reject(new Error('Missing payload url for sending to users device'))
      }

      nets({
        uri: 'https://chasqui.uport.me/api/v1/sns',
        json: {url},
        method: 'POST',
        withCredentials: false,
        headers: {
          Authorization: `Bearer ${token}`
        }
      },
      (error, res, body) => {
        if (error) return reject(error)
        if (res.statusCode === 200) {
          resolve(body)
        }
        if (res.statusCode === 403) {
          return reject(new Error('Error sending push notification to user: Invalid Token'))
        }
        reject(new Error(`Error sending push notification to user: ${res.statusCode} ${body.toString()}`))
      })
    })
  }

  // Create attestation
  attest ({sub, claim, exp}) {
    const subMNID = isMNID(sub) ? sub : encode({network: this.net.id, address: sub})
    return createJWT(this.settings, {sub: subMNID, claim, exp})
  }

  // Lookup public uport address of any user
  lookup (address) {
    return this.settings.registry(address)
  }
}
