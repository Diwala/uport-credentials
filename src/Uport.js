import { createJWT, verifyJWT } from './JWT'
import UportRegistry from 'uport-registry'

const INFURA_ROPSTEN = 'https://ropsten.infura.io'
const UPORT_REGISTRY_ADDRESS = '0xb9C1598e24650437a3055F7f66AC1820c419a679'

export default class Uport {
  constructor (settings) {
    this.settings = settings
    if (!this.settings.registry) {
      const registry = new UportRegistry()
      // The only function we're really interested from uport registry is getAttributes
      this.settings.registry = registry.getAttributes.bind(registry)
    }
  }

  // Create request token
  request (payload) {
    return createJWT(this.settings, {...payload, type: 'shareReq'})
  }

  // Receive response token from user and return data to promise
  receive (token) {
    return verifyJWT(this.settings, token).then(({payload, profile}) => (
      {...profile, ...(payload.own || {})}
    ))
  }

  // Create attestation
  attest ({sub, claim, exp}) {
    return createJWT(this.settings, {sub, claim, exp})
  }

  // send push 
  pushTo (pushToken, data, callback) {

  }
}
