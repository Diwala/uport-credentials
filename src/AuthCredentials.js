import Credentials from './Credentials.js'
import Storage from './Storage.js'
import crypto from 'crypto'
import { createJWT } from './JWT'

// Generate random 18 byte hex string
const random  = () => {
  return new Promise((resolve, reject) => {
    crypto.randomBytes(18, (err, buf) => {
      if (err) reject(error)
      resolve(buf.toString('hex'))
    })
  })
}

class AuthCredentials extends Credentials {
  constructor (settings = {}) {
    super(settings)
    this.challengeStorage = settings.challengeStorage || new Storage('challenge')
    this.responseStorage = settings.responseStorage || new Storage('response')
    this.random = settings.random || random
  }

  // Create request token

  // TODO the callback doesn't include the pair id
  createRequest (params = {}) {
    const payload = {}
    if (params.requested) {
      payload.requested = params.requested
    }
    if (params.verified) {
      payload.verified = params.verified
    }
    if (params.notifications) {
      payload.permissions = ['notifications']
    }
    if (params.callbackUrl) {
      payload.callback = params.callbackUrl
    }

    return Promise.all([this.random(), this.random()]).then(rand => {
      payload.challenge = rand[0]
      payload.pairId = rand[1]
    }).then(() => this.challengeStorage.set(payload.pairId, payload.challenge)
    ).then(res => createJWT(this.settings, {...payload, type: 'shareReq'}))
  }


  //TODO  Make response optional? then it doesn't write the response to the server
  receive (token, response = '', callbackUrl = null) {
    // Verify sig and check challenge equivalence
    let credentials
    return super.receive(token, callbackUrl).then(res => {
      credentials = res
      return this.challengeStorage.get(credentials.pairId)
    }).then(val => {
      const challenge = val
      if (challenge === credentials.challenge) {
        this.challengeStorage.del(credentials.pairId)
        return this.responseStorage.set(credentials.pairId, response)
      }
      this.responseStorage.set(credentials.pairId, 'Error')
      throw new Error('Authentication Failed: receive() Challenge did not match')
    }).then(res => credentials)
  }

  authResponse(pairId) {
    return this.responseStorage.get(pairId).then(res => {
      if (res) {
        this.responseStorage.del(pairId)
      }
      return res
    })
  }
}

export default AuthCredentials
