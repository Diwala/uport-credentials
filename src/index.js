import Credentials from './Credentials'
import SimpleSigner from './SimpleSigner'
import Contract from './Contract'
import JWT from './JWT'

module.exports = { Credentials, SimpleSigner, Contract, JWT }
