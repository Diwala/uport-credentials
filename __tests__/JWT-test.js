import { createJWT, verifyJWT } from '../src/JWT'
import SimpleSigner from '../src/SimpleSigner'
import { SECP256K1Client, TokenVerifier, decodeToken } from 'jsontokens'
import MockDate from 'mockdate'
MockDate.set(1485321133996)

const privateKey = '278a5de700e29faae8e40e366ec5012b5ec63d36ec77e8a2417154cc1d25383f'
const publicKey = SECP256K1Client.privateKeyToPublicKey(privateKey)
const signer = SimpleSigner(privateKey)
const verifier = new TokenVerifier('ES256K', publicKey)
const profileA = {publicKey: `0x${publicKey}`, name: 'David Chaum'}
const registry = (address) => new Promise((resolve, reject) => resolve(address === '0x001122' ? profileA : null))

it('creates a valid JWT', () => {
  return createJWT({address: '0x001122', signer}, {requested: ['name', 'phone']}).then((jwt) => {
    return expect(verifier.verify(jwt)).toBeTruthy()
  })
})

it('creates a JWT with correct format', () => {
  return createJWT({address: '0x001122', signer}, {requested: ['name', 'phone']}).then((jwt) => {
    return expect(decodeToken(jwt)).toMatchSnapshot()
  })
})

const incomingJwt = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJyZXF1ZXN0ZWQiOlsibmFtZSIsInBob25lIl0sImlzcyI6IjB4MDAxMTIyIiwiaWF0IjoxNDg1MzIxMTMzOTk2fQ.zxGLQKo2WjgefrxEQWfwm_oago8Qr4YctBJoqNAm2XKE-48bADjolSo2T_tED9LnSikxqFIM9gNGpNgcY8JPdg'

it('verifies the JWT and return correct payload', () => {
  return verifyJWT({registry, address: '0x001122'}, incomingJwt).then(({payload}) => {
    return expect(payload).toMatchSnapshot()
  })
})

it('verifies the JWT and return correct profile', () => {
  return verifyJWT({registry, address: '0x001122'}, incomingJwt).then(({profile}) => {
    return expect(profile).toEqual(profileA)
  })
})

it('rejects a JWT with missing publicKey', () => {
  return createJWT({address: '0x001123', signer}).then(jwt =>
    verifyJWT({registry, address: '0x001123'}, jwt).catch(error =>
      expect(error.message).toEqual('No profile found, unable to verify JWT')
    ).then((p) => expect(p).toBeFalsy())
  )
})

const badJwt = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NksifQ.eyJyZXF1ZXN0ZWQiOlsibmFtZSIsInBob25lIl0sImlzcyI6IjB4MDAxMTIyIiwiaWF0IjoxNDg1MzIxMTMzOTk2fQ.zxGLQKo2WjgefrxEQWfwm_oago8Qr4YctBJoqNAm2XKE-48bADjolSo2T_tED9LnSikxqFIM9gNGpNgcY8JPdf'
it('rejects a JWT with bad signature', () => {
  return verifyJWT({registry, address: '0x001122'}, badJwt).catch(error =>
    expect(error.message).toEqual('Signature invalid for JWT')
  ).then((p) => expect(p).toBeFalsy())
})

it('accepts a valid exp', () => {
  return createJWT({address: '0x001122', signer}, {exp: 1485321133996+1}).then(jwt =>
    verifyJWT({registry, address: '0x001122'}, jwt).then(({payload}) =>
      expect(payload).toMatchSnapshot()
    )
  )
})

it('rejects an expired JWT', () => {
  return createJWT({address: '0x001122', signer}, {exp: 1485321133996 - 1}).then(jwt =>
    verifyJWT({registry, address: '0x001122'}, jwt).catch(error =>
      expect(error.message).toEqual('JWT has expired')
    ).then((p) => expect(p).toBeFalsy())
  )
})

it('accepts a valid audience', () => {
  return createJWT({address: '0x001122', signer}, {aud: '0x001122'}).then(jwt =>
    verifyJWT({registry, address: '0x001122'}, jwt).then(({payload}) =>
      expect(payload).toMatchSnapshot()
    )
  )
})

it('accepts a valid audience using callback_url', () => {
  return createJWT({ address: '0x001122', signer }, { aud: 'http://chasqui.uport.me/unique' }).then(jwt =>
    verifyJWT({ registry }, jwt, 'http://chasqui.uport.me/unique').then(({payload}) =>
      expect(payload).toMatchSnapshot()
    )
  )
})

it('rejects invalid audience', () => {
  return createJWT({address: '0x001122', signer}, {aud: '0x001123' }).then(jwt =>
    verifyJWT({registry, address: '0x001122'}, jwt).catch(error =>
      expect(error.message).toEqual('JWT audience does not match your address')
    ).then((p) => expect(p).toBeFalsy())
  )
})
