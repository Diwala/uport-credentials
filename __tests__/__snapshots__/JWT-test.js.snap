exports[`test accepts a valid audience 1`] = `
Object {
  "aud": "0x001122",
  "iat": 1485321133996,
  "iss": "0x001122",
}
`;

exports[`test accepts a valid audience using callback_url 1`] = `
Object {
  "aud": "http://pututu.uport.me/unique",
  "iat": 1485321133996,
  "iss": "0x001122",
}
`;

exports[`test accepts a valid exp 1`] = `
Object {
  "exp": 1485321133997,
  "iat": 1485321133996,
  "iss": "0x001122",
}
`;

exports[`test creates a JWT with correct format 1`] = `
Object {
  "header": Object {
    "alg": "ES256K",
    "typ": "JWT",
  },
  "payload": Object {
    "iat": 1485321133996,
    "iss": "0x001122",
    "requested": Array [
      "name",
      "phone",
    ],
  },
  "signature": "zxGLQKo2WjgefrxEQWfwm_oago8Qr4YctBJoqNAm2XKE-48bADjolSo2T_tED9LnSikxqFIM9gNGpNgcY8JPdg",
}
`;

exports[`test verifies the JWT and return correct payload 1`] = `
Object {
  "iat": 1485321133996,
  "iss": "0x001122",
  "requested": Array [
    "name",
    "phone",
  ],
}
`;
